### ISSUE TITLE
False positive 'stop-iteration-return' when iterator is ensured to be non-empty

### ISSUE THREAD

**Issue by tuukkamustonen** [PyCQA/pylint#3424]
### Steps to reproduce

With:

```python
def foo(iterable):
    if not iterable:
        return
    it = iter(iterable)
    yield next(it)
```

### Current behavior

```
foo.py:5:10: R1708: Do not raise StopIteration in generator, use return statement instead (stop-iteration-return)
```

### Expected behavior

Isn't this false positive, as `iterable` was already ensured to contain element(s)?

### pylint --version output

```
pylint 2.4.4
astroid 2.3.1
Python 3.7.2 (default, Jan 29 2019, 14:27:19) 
[GCC 7.3.0]
```

**Comment by PCManticore** [comment]
Hey @tuukkamustonen Thanks for the report. 

Unfortunately this is an instance of the control flow series of issues we currently have in `pylint`. pylint does not take in consideration guards such as these, so regardless if a variable is protected against being `None`, empty or of a particular type with an `if` clause, most of the checks will still be emitted.

It's one of the issues that plagues pylint the most and we don't have a solution for it at this point. It will require adding support for flow inference, followed by adapting all the checks to respect the guards. But until that happens, there's not much you can do, other than disabling the check locally where it's emitted.

I'm going to close this one as it's just another datapoint for the ~30 issues we already have for control flow.

**Comment by tuukkamustonen** [comment]
Thanks for taking time to explain it even with already ~30 issues of the same topic!
